File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        int preIdx = 0;
        int inoIdx = 0;
        return BuildTree(preorder, inorder, ref preIdx, ref inoIdx, int.MaxValue);
    }

    private TreeNode BuildTree(int[] preorder, int[] inorder, ref int preIdx, ref int inoIdx, int stop) {
        if (preIdx >= preorder.Length) return null;
        if (inorder[inoIdx] == stop) {
            inoIdx++;
            return null;
        }

        TreeNode root = new TreeNode(preorder[preIdx]);
        preIdx++;
        root.left = BuildTree(preorder, inorder, ref preIdx, ref inoIdx, root.val);
        root.right = BuildTree(preorder, inorder, ref preIdx, ref inoIdx, stop);
        return root;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int result = std::numeric_limits<int>::min();
        dfs(root, result);
        return result;
    }

    int dfs(TreeNode* root, int& result) {
        if (root == nullptr) return 0;

        int left = std::max(0, dfs(root->left, result));
        int right = std::max(0, dfs(root->right, result));

        result = std::max(result, root->val + left + right);
        return root->val + std::max(left, right);
    }
};
